core/state/statedb_utils.go

```diff
@@ -93,16 +93,17 @@ func GetCandidates(statedb *StateDB) []common.Address {
 	slot := slotValidatorMapping["candidates"]
 	slotHash := common.BigToHash(new(big.Int).SetUint64(slot))
 	arrLength := statedb.GetState(common.HexToAddress(common.MasternodeVotingSMC), slotHash)
-	keys := []common.Hash{}
-	for i := uint64(0); i < arrLength.Big().Uint64(); i++ {
+	count := arrLength.Big().Uint64()
+	rets := make([]common.Address, 0, count)
+
+	for i := uint64(0); i < count; i++ {
 		key := GetLocDynamicArrAtElement(slotHash, i, 1)
-		keys = append(keys, key)
-	}
-	rets := []common.Address{}
-	for _, key := range keys {
 		ret := statedb.GetState(common.HexToAddress(common.MasternodeVotingSMC), key)
-		rets = append(rets, common.HexToAddress(ret.Hex()))
+		if !ret.IsZero() {
```

Comment:
If I'm correct, the reason that we have to check `IsZero()` is that the smart contract returns zero address which is a bug inside the smart contract. It is because the smart contract has deleted some address but not "shrink" the array to hold the address. Good job to catch it anyway!

Reply:
In the solidity language, delete a[x] deletes the item at index x of the array and leaves all other elements and the length of the array untouched. This especially means that it leaves a gap in the array.  So in the contract `XDCValidator`, the command `delete candidates[i]` only set `candidates[i]` to zero value, this creates zero address in the array candidates.